returnn/tf/engine.py

```diff
@@ -2316,27 +2320,80 @@ def extra_fetches_callback(seq_idx, seq_tag, **kwargs):
+        if beam_scores[output_layer_idx] is not None:
+          assert beam_scores[output_layer_idx].shape == (n_batch, out_beam_sizes[output_layer_idx])
 
-        assert n_batch * (out_beam_sizes[target_idx] or 1) == len(outputs[target_idx])
-        if do_eval and targets[target_idx] is not None:
-          assert n_batch == len(targets[target_idx])
+        assert n_batch * (out_beam_sizes[output_layer_idx] or 1) == len(outputs[output_layer_idx])
+        if do_eval and targets[output_layer_idx] is not None:
+          assert n_batch == len(targets[output_layer_idx])
 
-        if output_layers[target_idx].output.dim == 256 and output_layers[target_idx].output.sparse:
+        if output_layers[output_layer_idx].output.dim == 256 and output_layers[output_layer_idx].output.sparse:
           # Interpret output as bytes/utf8-string.
-          outputs[target_idx] = bytearray(outputs[target_idx]).decode("utf8")
+          outputs[output_layer_idx] = bytearray(outputs[output_layer_idx]).decode("utf8")
+
+      # Create lists with serialized data. All of length num_output_layers.
+      serialized_outputs = []  # type: typing.List[typing.Optional[typing.Union[str,numpy.ndarray]]]
+      serialized_targets = []  # type: typing.List[typing.Optional[typing.Union[str,numpy.ndarray]]]
+      # noinspection PyShadowingNames
+      for output_layer_idx in range(num_output_layers):
+        if output_layers[output_layer_idx].output.sparse:
+          # Try to get the vocab corresponding to the sparse dim
+          vocab = None
+          if output_layers[output_layer_idx].output.sparse_dim.vocab is not None:
+            vocab = output_layers[output_layer_idx].output.sparse_dim.vocab
+          elif target_keys[output_layer_idx] and (
+            self.network.extern_data.get_data(target_keys[output_layer_idx]).vocab is not None):
+            vocab = self.network.get_extern_data(target_keys[output_layer_idx]).vocab
+          elif target_keys[output_layer_idx] and target_keys[output_layer_idx] in dataset.labels:
+            vocab = Vocabulary.create_vocab_from_labels(dataset.labels[target_keys[output_layer_idx]])
+
+          if vocab is not None:
+            layer_outputs = outputs[output_layer_idx]
+            serialized_output = [
+              vocab.get_seq_labels(output) if output.ndim == 1 else vocab.labels[output] for output in layer_outputs]
+          else:
+            serialized_output = None
+            assert not output_file, "Unable to serialize sparse output of layer '%s'." % (
+              output_layer_names[output_layer_idx])
+        else:
+          # Output dense layers as-is
+          serialized_output = outputs[output_layer_idx]
+        serialized_outputs.append(serialized_output)
+
+        if target_keys[output_layer_idx] and do_eval:
+          target_extern_data = self.network.get_extern_data(target_keys[output_layer_idx])
+          if not target_extern_data.sparse:
```

Comment:
Wrong way around, right?
```suggestion
          if target_extern_data.sparse:
```